include/boost/json/detail/impl/stack.ipp

```diff
@@ -35,7 +35,8 @@ reserve(std::size_t n)
     if(buf_ && size_ > 0)
         std::memcpy(
             buf, buf_, size_);
-    sp_->deallocate(buf_, size_);
+    if(buf_)
+        sp_->deallocate(buf_, cap_);
```

Comment:
Consider folding the condition?
```
    if( buf_ )
    {
        if( size_ > 0 )
            std::memcpy( buf, buf_, size_ );
        sp_->deallocate(buf_, size_);
    }
```

Reply:
Or even skip the check for buf_, replacing it with the size check and an assert on buf_?

Reply:
The idea being that if size is >0, buf_ should definitely be nonnull or else something has gone haywire somewhere else.

Reply:
That may be possible, but you would have to look at the entire class to know if that precondition is correct. If you do that, use `BOOST_ASSERT`. However, just a guess but it might not be possible, as `size_` can be zero while `buf_` is nonzero. This is the case when the stack grows, and then everything is popped off, leaving it with a non-zero capacity but a zero size.

Reply:
I will go for the folded condition :-)